### ISSUE TITLE
ModelValueResolver resolves same argument multiple times

### ISSUE THREAD

**Issue by RobinHoutevelts** [wieni/wmmodel#21]
## Description

The `ModelValueResolver` resolves the same argument multiple times in some cases.

### Steps to reproduce

For instance with this controller:

```php
class Controller {
    public function show(Page $a, Page $b) {}
}
```

and this route:

```yaml
my_route:
    path: 'foobar/{a}/{b}'
    defaults:
        _controller: 'Controller::show'
    options:
        parameters:
            a: {type: 'entity:node'}
            b: {type: 'entity:node'}
```

And then visiting `/foobar/123/456`

### Expected result

I expect that when the `show` method of my controller is called, the`$b` argument is resolved to a Node with id `456`.

### Actual result

The `$a` and `$b` argument are the same.

**Comment by RobinHoutevelts** [comment]
It's easy to test for. Here is a failing test:

```
$parent = $this->createMock(User::class);
$child = $this->createMock(User::class);

$request = new Request();
$request->attributes->set('a', $parent);
$request->attributes->set('b', $child);

$controllerClass = new class {
    public function show(User $a, User $b) {}
};

$controller = [$controllerClass, 'show'];
$resolver = new ArgumentResolver(argumentValueResolvers: [new ModelValueResolver()]);

$arguments = $resolver->getArguments($request, $controller);

static::assertSame($parent, $arguments[0], 'The $a variable is not resolved to $parent');
static::assertSame($child, $arguments[1], 'The $b variable is not resolved to $child');
```

**Comment by RobinHoutevelts** [comment]
In https://github.com/wieni/wmmodel/commit/9c2671ac6a8180b1ea8c8081101dda1c505db23d#diff-e8a4c73a044ab3e1c57d89244a70d3ca1eda94fa4c2987c61234d626f1e43e45L101 we created the ModelValueResolver but didn't take this into account.

Originally we did: https://github.com/wieni/wmmodel/blob/0.3.7/src/Controller/ControllerResolver.php#L116

**Comment by DieterHolvoet** [comment]
We should probably compare parameter names with argument names, and only fallback to 'any object of the same type' if the names don't match. To make it more robust we could compare different versions of the names, camel case and snake case.

**Comment by RobinHoutevelts** [comment]
Exactly. The test passes when we do this:

```php

public function resolve(Request $request, ArgumentMetadata $argument)
{
    // First look for an exact match based on type AND name
    if ($attribute = $request->attributes->get($argument->getName())) {
        if ($this->isMatch($attribute, $argument)) {
            yield $attribute;

            return;
        }
    }

    foreach ($request->attributes->getIterator() as $name => $attribute) {
        if ($this->isMatch($attribute, $argument)) {
            yield $attribute;

            return;
        }
    }

    if ($argument->hasDefaultValue()) {
        yield $argument->getDefaultValue();

        return;
    }
}

private function isMatch($attribute, ArgumentMetadata $argument): bool
{
    return is_object($attribute) && is_a($attribute, $argument->getType());
}
```

**Comment by HnLn** [comment]
Above code in https://github.com/wieni/wmmodel/pull/23

**Comment by DieterHolvoet** [comment]
> I would avoid adding this kind of logic again.

Don't think we ever had that logic. I vaguely remember some pluralizing going on, but not this. 

>  I don't see any use case for it.

The convention in routes is to use snake case for parameter names, and we like to use camel case for our argument names. That's the use case. Also exactly why we have https://github.com/wieni/wmmodel/issues/18.

eg.
```yaml
entity.taxonomy_term.canonical:
  path: '/taxonomy/term/{taxonomy_term}'
  defaults:
    _entity_view: 'taxonomy_term.full'
    _title: 'Taxonomy term'
    _title_callback: '\Drupal\taxonomy\Controller\TaxonomyController::termTitle'
  requirements:
    _entity_access: 'taxonomy_term.view'
    taxonomy_term: \d+
```

**Comment by DieterHolvoet** [comment]
Pushed a change to the MR, wasn't able to test it though since I didn't immediately have a project with wmmodel running. Let me know if it works.

**Comment by RobinHoutevelts** [comment]
I meant it's the same _kind_ of logic/magic that I would want to avoid.

Your example doesn't fit because you only pass `taxonomy_term` which would be resolved by typehinting. Like it is now.

#18 also doesn't fit. Unless you have multiple FormState arguments in your controller method. Which afaik never happens.

**Comment by DieterHolvoet** [comment]
> Your example doesn't fit

I know, that was just an example for the casing I copy pasted. Here's your example updated:
```yaml
my_route:
    path: 'foobar/{some_param}/{other_param}'
    defaults:
        _controller: 'Controller::show'
    options:
        parameters:
            some_param: {type: 'entity:node'}
            other_param: {type: 'entity:node'}
```
```php
class Controller {
    public function show(Page $someParam, Page $otherParam) {}
}
```

**Comment by RobinHoutevelts** [comment]
Ok. So just rename your route arguments? Instead of relying on us to magically make that work.

**Comment by DieterHolvoet** [comment]
Core is going to support form_state and formState ([issue](https://www.drupal.org/project/drupal/issues/3006502)), but they can do it through type resolving since there's only one form state like you said. It just seems useful to me to do the same, only we have to do a more complicated check.

**Comment by DieterHolvoet** [comment]
> It's fine. If you think it's a useful feature for wmmodel I'm willing to go with it

Okay, thanks 👍

**Comment by HnLn** [comment]
The wmcontent edit paths seem to work again, including dieter's commit (https://github.com/wieni/wmcontent/blob/main/src/Routing/WmContentRouteSubscriber.php#L162).  So can be considered as tested.

**Comment by DieterHolvoet** [comment]
Is it also tested in this kind of situation? I can do it but only if I find the time, probably somewhere during the next week.
```yaml
my_route:
    path: 'foobar/{some_param}/{other_param}'
    defaults:
        _controller: 'Controller::show'
    options:
        parameters:
            some_param: {type: 'entity:node'}
            other_param: {type: 'entity:node'}
```
```php
class Controller {
    public function show(Page $someParam, Page $otherParam) {}
}
```

**Comment by HnLn** [comment]
nope only: /node/{node}/wmcontent/{container}/{child}/edit

with https://github.com/wieni/wmcontent/blob/main/src/Controller/WmContentChildController.php#L112

(and the edit callback as wel, as before it returned the host as child instead of the content block)

**Comment by RobinHoutevelts** [comment]
I've tested that scenario:

```php
$someParam = $this->createMock(User::class);
$otherParam = $this->createMock(User::class);

$request = new Request();
$request->attributes->set('some_param', $someParam);
$request->attributes->set('other_param', $otherParam);

$controllerClass = new class {
    public function show(User $otherParam, User $someParam)
    {
    }
};

$controller = [$controllerClass, 'show'];
$resolver = new ArgumentResolver(argumentValueResolvers: [new ModelValueResolver()]);

$arguments = $resolver->getArguments($request, $controller);

static::assertSame($otherParam, $arguments[0], 'The first variable is not resolved to $otherParam');
static::assertSame($someParam, $arguments[1], 'The second variable is not resolved to $someParam');
```

All is green 👍 

<img width="287" alt="Screenshot 2022-01-20 at 15 21 02" src="https://user-images.githubusercontent.com/9056689/150356405-7cb16461-2fc5-4c39-abee-d42e926fa88d.png">

**Comment by DieterHolvoet** [comment]
Great, thanks! Can you maybe add those tests to the repo? There are already a few tests present.

**Comment by RobinHoutevelts** [comment]
Closed in #23
